### PR TITLE
Add async_set_datetime and async_get_datetime methods

### LINKED ISSUE #28: Set time and date via SNMP
On my DCP-J552DW each time there is power outage I need to setup time and date for scan uploads. Is it possible to do this via SNMP and to add it here? It could be a nice feature to recover time and date each time it recovers from HA.

### PR DESCRIPTION
**Pull request by lorek123**
Closes #28

## Summary

Many Brother printers (especially older inkjet models like the DCP-J552DW) lose their clock after a power outage, which breaks scanning functionality until the date/time is manually reset on the control panel.

This PR adds SNMP-based methods to **read and set the printer's date/time** via the standard `hrSystemDate.0` OID (`1.3.6.1.2.1.25.1.2.0`), enabling automated clock restoration (e.g. via Home Assistant).

### New API

- **`Brother(write_community=...)`** — optional parameter for the SNMP write community string (defaults to `"internal"`, which is the standard Brother write community)
- **`async_get_datetime() -> datetime | None`** — reads the printer's current clock as a naive datetime (printer has no timezone concept)
- **`async_set_datetime(dt=None)`** — sets the printer's clock; defaults to the host's local time if no datetime is provided

### Changes

- `brother/__init__.py` — added `set_cmd` import, `write_community` parameter, `async_get_datetime`, `async_set_datetime`, `_build_dateandtime`, `_parse_dateandtime`
- `brother/const.py` — added `OID_DATETIME`, `DATEANDTIME_MIN_LENGTH`, `DEFAULT_WRITE_COMMUNITY` constants
- `tests/test_datetime.py` — 14 new tests covering set/get, error paths, DateAndTime encoding/decoding, and community string configuration

### Background

Tested live on a Brother DCP-J552DW. The `hrSystemDate.0` OID is a standard HOST-RESOURCES-MIB OID supported across Brother models. The write community `"internal"` is the default Brother SNMP write community string, discovered via SNMP reconnaissance and firmware analysis of `BroSNMP.dll`.

## Test plan

- [x] All 70 tests pass (14 new + 56 existing, 0 regressions)
- [x] `ruff check` passes with no errors
- [x] Verified `async_set_datetime()` and `async_get_datetime()` against a real DCP-J552DW printer
- [x] Confirmed correct local timezone handling (tested from CET/Poland timezone)

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * View the device's current date and time.
  * Set the device's date and time, including an option to use the system time.
  * Configure a custom SNMP write community for write operations.

* **Tests**
  * Added comprehensive unit tests covering datetime get/set flows, error mappings, and encoding/decoding edge cases.

### CODE REVIEW

File: tests/test_datetime.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+"""Tests for datetime get/set functionality."""
```
Comment (bieniu) [inline tests/test_datetime.py]: The test file should match the code file. There is no datetime.py file in the Brother library.

File: brother/__init__.py [hunk 2]
```diff
@@ -282,6 +302,112 @@ def shutdown(self) -> None:
...
+    @staticmethod
+    def _build_dateandtime(dt: datetime) -> bytes:
+        """Encode a datetime as an 8-byte SNMP DateAndTime value (RFC 2579)."""
+        return (
+            dt.year.to_bytes(2, "big")
+            + bytes([dt.month, dt.day, dt.hour, dt.minute, dt.second, 0])
+        )
+
+    @staticmethod
+    def _parse_dateandtime(raw: bytes) -> datetime | None:
```
Comment (bieniu) [inline brother/__init__.py]: This function should be moved to the `utils` module.

File: brother/__init__.py [hunk 3]
```diff
@@ -282,6 +302,112 @@ def shutdown(self) -> None:
...
+        """Return SNMP request args with the given community string."""
+        return (
+            self._request_args[0],
+            CommunityData(community, mpModel=0),
+            self._request_args[2],
+            self._request_args[3],
+        )
+
+    @staticmethod
+    def _build_dateandtime(dt: datetime) -> bytes:
```
Comment (bieniu) [inline brother/__init__.py]: This function should be moved to the `utils` module.

File: tests/test_datetime.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import pytest
+from pysnmp.error import PySnmpError
+from pysnmp.hlapi.v3arch.asyncio import CommunityData, ContextData, SnmpEngine
+
+from brother import Brother, SnmpError
+
+HOST = "localhost"
+
+
+def _setup_request_args(brother: Brother) -> None:
```
Comment (bieniu) [inline tests/test_datetime.py]: This should be a fixture.

**Review by bieniu — CHANGES_REQUESTED** [review]
Thanks for your PR. I didn't have time to thoroughly review the code, so here are a few comments that caught my eye.

**Comment by lorek123** [comment]
@bieniu I think all requested changes are now solved, please take a look

**Comment by bieniu** [comment]
The code looks good.
Unfortunately, setting the time doesn't work for my printer. SNMP doesn't return an error, but the time isn't set and the printer always returns `2000-01-01 00:00:00`.

Therefore, I'd like to ask how you plan to use these two new methods. Do you intend to implement the Brother integration change in the HA repository? If so, I'd like you to open a draft PR so we can analyze how it will work.

**Comment by lorek123** [comment]
@bieniu I can work on broader support, I'll just need few more details. If you want we can catch up on Discord: lorek123.

File: brother/__init__.py [hunk 5]
```diff
@@ -282,6 +304,87 @@ def shutdown(self) -> None:
         if self._snmp_engine:
             LCD.unconfigure(self._snmp_engine, None)
 
+    async def async_get_datetime(self) -> datetime | None:
+        """Read the printer's current date and time via SNMP.
```
Comment (bieniu) [inline brother/__init__.py]: We don't need such an extended docstring here.

File: brother/__init__.py [hunk 6]
```diff
@@ -282,6 +304,87 @@ def shutdown(self) -> None:
...
+            if str(errstatus) == "noSuchName":
+                return None
+            msg = f"SNMP GET failed: {errstatus} at index {errindex}"
+            raise SnmpError(msg)
+
+        raw: bytes = restable[0][-1].asOctets()
+        return parse_dateandtime(raw)
+
+    async def async_set_datetime(self, dt: datetime | None = None) -> None:
+        """Set the printer's date and time via SNMP.
```
Comment (bieniu) [inline brother/__init__.py]: We don't need such an extended docstring here.

File: brother/__init__.py [hunk 7]
```diff
@@ -282,6 +304,87 @@ def shutdown(self) -> None:
...
+
+        if errindication:
+            raise SnmpError(str(errindication))
+        if errstatus:
+            msg = f"SNMP SET failed: {errstatus} at index {errindex}"
+            raise SnmpError(msg)
+
+        _LOGGER.debug("Printer datetime set to %s", dt.isoformat())
+
+    def _request_args_for(
```
Comment (bieniu) [inline brother/__init__.py]: The function name is misleading.

File: brother/__init__.py [hunk 8]
```diff
@@ -282,6 +304,87 @@ def shutdown(self) -> None:
...
+        if errstatus:
+            if str(errstatus) == "noSuchName":
+                return None
+            msg = f"SNMP GET failed: {errstatus} at index {errindex}"
+            raise SnmpError(msg)
+
+        raw: bytes = restable[0][-1].asOctets()
+        return parse_dateandtime(raw)
+
+    async def async_set_datetime(self, dt: datetime | None = None) -> None:
```
Comment (bieniu) [inline brother/__init__.py:333]: We know that setting the date doesn't work on all printer models. Therefore, we need to maintain a list of printers that support this feature and raise an exception when attempting to use this method on a printer model that isn't on the list.

File: brother/__init__.py [hunk 9]
```diff
@@ -69,6 +78,7 @@ def __init__(
         printer_type: str = "laser",
         model: str | None = None,
         snmp_engine: SnmpEngine | None = None,
+        write_community: str | None = None,
```
Comment (bieniu) [inline brother/__init__.py]:
```suggestion
        write_community: str = DEFAULT_WRITE_COMMUNITY,
```

**Review by bieniu — CHANGES_REQUESTED** [review]

File: brother/__init__.py (hunk 8 above)
Comment (lorek123) [reply brother/__init__.py:333]: Added an allowlist, do we want to have a way to test for users if their printer has this option?

Comment (bieniu) [reply brother/__init__.py:333]: yeah, good idea

File: brother/__init__.py [hunk 10]
```diff
@@ -282,6 +303,72 @@ def shutdown(self) -> None:
...
+        return parse_dateandtime(raw)
+
+    async def async_set_datetime(self, dt: datetime | None = None) -> None:
+        """Set the printer's date and time via SNMP."""
+        model = getattr(self, "model", None)
+        if model is None or not any(
+            m in model.lower() for m in DATETIME_SET_SUPPORTED_MODELS
+        ):
+            msg = f"Setting datetime is not supported on model {model}"
+            raise UnsupportedModelError(msg)
```
Comment (bieniu) [inline brother/__init__.py]: Let's add a new exception for this purpose, e.g. `MethodNotSupportedError`. `UnsupportedModelError` doesn't suit here

https://github.com/bieniu/brother/blob/b5430b8e5f11dbda17c6328932d2a7a4107faa39/brother/exceptions.py#L17-L18

File: brother/__init__.py [hunk 11]
```diff
@@ -282,6 +303,72 @@ def shutdown(self) -> None:
...
+                return None
+            msg = f"SNMP GET failed: {errstatus} at index {errindex}"
+            raise SnmpError(msg)
+
+        raw: bytes = restable[0][-1].asOctets()
+        return parse_dateandtime(raw)
+
+    async def async_set_datetime(self, dt: datetime | None = None) -> None:
+        """Set the printer's date and time via SNMP."""
+        model = getattr(self, "model", None)
```
Comment (bieniu) [inline brother/__init__.py]: The printer model is available as self.model, we don't need to use `getattr()`

File: brother/__init__.py [hunk 12]
```diff
@@ -282,6 +303,72 @@ def shutdown(self) -> None:
...
+        if model is None or not any(
```
Comment (bieniu) [inline brother/__init__.py]: `self.model` is a string, no need to check if it is `None`

**Review by bieniu — CHANGES_REQUESTED** [review]

**Review by bieniu — APPROVED** [review]
Looks good, thanks @lorek123 👍
